runtime/runtime/src/metrics.rs

```diff
@@ -558,3 +616,54 @@ impl ApplyMetrics {
             .observe(self.accumulated_compute as f64 / TERA);
     }
 }
+
+pub fn report_congestion_metrics(receipt_sink: &ReceiptSink, sender_shard_id: ShardId) {
+    match receipt_sink {
+        ReceiptSink::V1(_) => {
+            // no metrics to report
+        }
+        ReceiptSink::V2(inner) => {
+            let sender_shard_label = sender_shard_id.to_string();
+            report_congestion_indicators(&inner.congestion_info, &sender_shard_label);
+            report_outgoing_buffers(inner, sender_shard_label);
+        }
+    }
+}
+
+/// Report key congestion indicator levels of a shard.
+fn report_congestion_indicators(congestion_info: &CongestionInfo, shard_label: &str) {
+    let congestion_level = congestion_info.congestion_level();
+    CONGESTION_LEVEL.with_label_values(&[shard_label]).set(congestion_level);
+
+    let CongestionInfo::V1(inner) = congestion_info;
+    CONGESTION_RECEIPT_BYTES
+        .with_label_values(&[shard_label])
+        .set(inner.receipt_bytes.try_into().unwrap_or(i64::MAX));
+    CONGESTION_INCOMING_GAS
+        .with_label_values(&[shard_label])
+        .set(inner.delayed_receipts_gas.try_into().unwrap_or(i64::MAX));
+    CONGESTION_OUTGOING_GAS
+        .with_label_values(&[shard_label])
+        .set(inner.buffered_receipts_gas.try_into().unwrap_or(i64::MAX));
+}
+
+/// From `sender_shard` to all other shards, reports how many receipts are
+/// currently buffered and how much forwarding capacity was left.
+fn report_outgoing_buffers(
+    inner: &crate::congestion_control::ReceiptSinkV2,
+    sender_shard_label: String,
+) {
+    for (&receiver_shard_id, &unused_capacity) in inner.outgoing_limit.iter() {
+        let receiver_shard_label = receiver_shard_id.to_string();
+
+        RECEIPT_FORWARDING_UNUSED_CAPACITY_GAS
+            .with_label_values(&[&sender_shard_label, &receiver_shard_label])
+            .set(i64::try_from(unused_capacity).unwrap_or(i64::MAX));
```

Comment:
I don't think this'll work well with round robin. This metric will be set on every block but prometheus queries for the metrics only every so often. It's still valuable for the intermediate stages though so I'm ok having it as is.

Reply:
Good point, I haven't thought about round-robin.

But I think it's fine. Whether we see 0 or 1 Pgas, both is low enough that we know there is backpressure. Also, unused capacity should be flat at zero all the time during round-robin, unless we have nothing to forward, which is easy to see in `near_congestion_outgoing_receipt_buffer_len`.